File: python/Q1448_count_good_nodes_in_binary_tree.py
#
# @lc app=leetcode id=1448 lang=python3
#
# [1448] Count Good Nodes in Binary Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def good_nodes(self, root: TreeNode, maximum: int) -> int:
        result: int = 0
        maximum_value: int = maximum if maximum > root.val else root.val

        if root.left:
            result += self.good_nodes(root.left, maximum_value)

        if root.right:
            result += self.good_nodes(root.right, maximum_value)

        if root.val >= maximum:
            result += 1

        return result

    def goodNodes(self, root: TreeNode) -> int:
        return self.good_nodes(root, -(10**4) - 1)


# @lc code=end

File: rust/src/leetcode/Q1448_count_good_nodes_in_binary_tree.rs
/*
 * @lc app=leetcode id=1448 lang=rust
 *
 * [1448] Count Good Nodes in Binary Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    fn count_good_nodes_with_cap(root: Option<Rc<RefCell<TreeNode>>>, mut cap: i32) -> i32 {
        if let Some(node) = root {
            let mut result: i32 = 0;

            if node.borrow().val >= cap {
                result += 1;
                cap = node.borrow().val;
            }

            result += Self::count_good_nodes_with_cap(node.borrow().left.clone(), cap);
            result += Self::count_good_nodes_with_cap(node.borrow().right.clone(), cap);

            result
        } else {
            0
        }
    }

    pub fn good_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::count_good_nodes_with_cap(root, i32::MIN)
    }
}
// @lc code=end
